### PR TITLE
arch: fix cmake's ARM CRC intrinsics test to handle duplicitous gcc 4.8.5

### PR DESCRIPTION
**Pull request by dmick**

### CODE REVIEW

**Comment by dmick** [comment]
Actual test build here:

https://jenkins.ceph.com/job/ceph-dev-new-build/ARCH=arm64,AVAILABLE_ARCH=arm64,AVAILABLE_DIST=centos7,DIST=centos7,MACHINE_SIZE=huge/2037/consoleFull

When I looked, it had the lack of intrinsics:

-- Performing Test HAVE_ARMV8_CRC_CRYPTO_INTRINSICS - Failed

and successfully built the crc file as a result:

Building C object src/CMakeFiles/common_crc_aarch64.dir/common/crc32c_aarch64.c.o
Linking C static library ../lib/libcommon_crc_aarch64.a
[  1%] Built target common_crc_aarch64

File: cmake/modules/SIMDExt.cmake [hunk 1]
```diff
@@ -34,28 +35,44 @@ if(CMAKE_SYSTEM_PROCESSOR MATCHES "aarch64|AARCH64")
...
+
+  # don't believe only the -march support; gcc 4.8.5 on RHEL/CentOS says
+  # it supports +crc but hasn't got the intrinsics or arm_acle.h.  Test for
+  # the actual presence of one of the intrinsic functions.
+  if(HAVE_ARMV8_CRC_CRYPTO_INTRINSICS_MARCH)
+    check_cxx_source_compiles("
+      #include <inttypes.h>
+      int main() { uint32_t a; uint8_t b; __builtin_aarch64_crc32b(a, b); }
+    " HAVE_ARMV8_CRC_CRYPTO_INTRINSICS)
+  endif()
```
Comment (tchaikov) [inline cmake/modules/SIMDExt.cmake]: makes sense to me.

Comment (tchaikov) [inline cmake/modules/SIMDExt.cmake]: as "-m<machine>" does not imply that the compiler supports the intrinsics.

Comment (dmick) [inline cmake/modules/SIMDExt.cmake]: oh...I guess that's true, huh; it only indicates that the compiler can output the instructions.  Hm.  Maybe this shouldn't even bother with -march

Comment (tchaikov) [inline cmake/modules/SIMDExt.cmake]: but do we need to detect the support of `-march=armv8-a+crc+crypto` before using the crc32 intrinsics ? IIUC, `-m<machine>` instructs the compiler backend to emit instructions offered by that `$machine`, but in this case, we are not passing `-march=armv8-a+crc+crypto` to achieve this goal, instead we manually optimize the crc32 digest algorithm. so i guess we can just drop `CHECK_C_COMPILER_FLAG(-march=armv8-a+crc+crypto HAVE_ARMV8_CRC_CRYPTO_INTRINSICS_MARCH)`, and detect `HAVE_ARMV8_CRC_CRYPTO_INTRINSICS` using `check_cxx_source_compiles(...)`.

Comment (dmick) [inline cmake/modules/SIMDExt.cmake]: Maybe?....I dunno what happens if you try to refer to the intrinsic but do not pass crc+crypto.

Comment (dmick) [inline cmake/modules/SIMDExt.cmake]: Answer: if you don't pass -march=armv8-a+crc, you get an error from assembly.

`$ cat test.c
#include <inttypes.h>
int main() { uint32_t a; uint8_t b; __builtin_aarch64_crc32b(a, b); }
$ gcc test.c -o test
test.c: In function ‘main’:
test.c:2:1: error: unrecognizable insn:
 int main() { uint32_t a; uint8_t b; __builtin_aarch64_crc32b(a, b); }
 ^
(insn 7 6 8 2 (set (reg:SI 75)
        (unspec:SI [
                (reg:SI 76)
                (reg:QI 77)
            ] UNSPEC_CRC32B)) test.c:2 -1
     (nil))
test.c:2:1: internal compiler error: in extract_insn, at recog.c:2343
$ gcc -march=armv8-a+crc test.c -o test
$

Comment (tchaikov) [inline cmake/modules/SIMDExt.cmake]: >  i dunno what happens if you try to refer to the intrinsic but do not pass crc+crypto.

the intrinsics are implemented using the the GCC builtin functions like `__builtin_aarch64_crc32b()` which is directly mapped to the corresponding assembly without the help of the `-m<machine>`. (i skimmed https://github.com/gcc-mirror/gcc/blob/master/gcc/config/aarch64/aarch64-builtins.c, and got this conclusion =))

>  so i guess we can just drop CHECK_C_COMPILER_FLAG(-march=armv8-a+crc+crypto HAVE_ARMV8_CRC_CRYPTO_INTRINSICS_MARCH)

okay, we cannot drop it, as we are trying to set `ARMV8_CRC_COMPILE_FLAGS` if `-march=armv8-a+crc+crypto` is supported. so maybe we can s/HAVE_ARMV8_CRC_CRYPTO_INTRINSICS/HAVE_ARMV8_CRC_CRYPTO_INTRINSICS_MARCH/` at line 59. and maybe rename `HAVE_ARMV8_CRC_CRYPTO_INTRINSICS_MARCH` to `HAVE_ARMV8_CRC_CRYPTO_MARCH`.

Comment (tchaikov) [inline cmake/modules/SIMDExt.cmake]: @dmick thanks for confirming this! should have tested it. but could you consider the second suggestion in the above comment?

Comment (dmick) [inline cmake/modules/SIMDExt.cmake]: I think it makes the most sense to use HAVE_ARMV8_CRC_CRYPTO_MARCH and HAVE_ARMV8_CRC_CRYPTO_INTRINSICS.  It's true we can't use the second if we don't have the  first, but we're testing for them separately.  But I agree it makes it better to have the first named more accurately.

**Comment by dmick** [comment]
Rebased version built.  I'll merge.

https://jenkins.ceph.com/job/ceph-dev-new-build/2085/console

**Pull request merged**